### PR TITLE
Initial java openliberty stack

### LINKED ISSUE #364: Update java-microprofile stack to match the best practice Open Liberty pom
We are aiming to simplify the maven pom.xml used for Open Liberty projects to something based on the snippet below. This needs some modification to fit into Appsody, but it is a much simpler approach to the current one in the appsody template for micro profile.

```
<?xml version='1.0' encoding='utf-8'?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>my.demo</groupId>
    <artifactId>demo-project</artifactId>
    <version>1.0-SNAPSHOT</version>
    <packaging>war</packaging>

    <properties>
        <maven.compiler.source>1.8</maven.compiler.source>
        <maven.compiler.target>1.8</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
    </properties>

    <dependencies>
        <dependency>
            <groupId>org.eclipse.microprofile</groupId>
            <artifactId>microprofile</artifactId>
            <version>3.0</version>
            <type>pom</type>
            <scope>provided</scope>
        </dependency>
    </dependencies>

    <build>
        <finalName>${project.artifactId}</finalName>
        <plugins>
            <plugin>
                <groupId>io.openliberty.tools</groupId>
                <artifactId>liberty-maven-plugin</artifactId>
                <version>3.0.1</version>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-war-plugin</artifactId>
                <version>3.2.2</version>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>2.22.2</version>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-failsafe-plugin</artifactId>
                <version>2.22.2</version>
            </plugin>
        </plugins>
    </build>
</project>
```

### LINKED ISSUE #584: Create a stack based on Open Liberty using liberty-maven-plugin v3.0 "dev mode"
**Is your feature request related to a problem? Please describe.**
The java-microprofile stack performance (e.g. startup time on `appsody run`) isn't good enough.

**Describe the solution you'd like**
Better performance on `appsody run` startup as well as upon app changes in local development mode.

**Describe alternatives you've considered**
Thought it would be conceivable to update the existing java-microprofile stack in this direction, we will probably make a breaking change to apps (since v3.x of liberty-maven-plugin uses a new **io.openliberty.tools** groupid, compared to v2.x).   In addition, the "java-openliberty" name better fits the capabilities of an Open Liberty-based runtime, since it includes Jakarta APIs and features not just MicroProfile features.      This seems like an opportunity to make a fresh start.

### PR DESCRIPTION
**Pull request by scottkurz**
### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stack-structure.md).

### Modifying an existing stack:

- N/A

### Contributing a new stack:

- Describe how application dependencies are managed:

This stack is Maven-based and this aspect functions largely like the java-microprofile predecessor.

The app pom.xml references the stack's pom.xml as its parent POM.    When launched, e.g. via `appsody run`, the container mounts the host .m2 repo to leverage the already-downloaded dependencies in the user's host environment.  

- Explain how Appsody file watcher is utilized:

We are currently only watching **pom.xml**, since lauching `mvn liberty:dev` launches a "watcher" with the liberty maven plugin (i.e. it launches in "dev mode").    (And this special case should probably be removed and folded into the others soon, using newer liberty-maven-plugin versions).

- Describe other Appsody environment variables defined by the stack image:

No special env vars.  It would help though to describe the way that maven profiles are used to differentiate between the "local dev" and "build/deploy" modes, as well as in the host-only (non-Docker) "mode" (which may not be required by Appsody but can still be useful).

- Describe any limitations and known issues:

See the [README.md](https://github.com/appsody/stacks/blob/master/incubator/java-openliberty/README.md) for more info (or [this link](https://github.com/scottkurz/stacks/blob/initial-java-openliberty-stack/incubator/java-openliberty/README.md) in the meantime until the PR is merged) especially the section on **Windows 10** issues.

### Related Issues:
Fixes #584, fixes #364

### CODE REVIEW

**Comment by neeraj-laad** [comment]
Does this PR replace #552 ?

**Comment by scottkurz** [comment]
> 
> 
> Does this PR replace #552 ?

Yes (closed that now).

File: incubator/java-openliberty/image/project/pom.xml [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                            <groupId>io.openliberty.tools</groupId>
+                            <artifactId>liberty-maven-plugin</artifactId>
+                            <version>${version.liberty-maven-plugin}</version>
+                            <configuration>
+                                <assemblyArtifact>
+                                    <groupId>${liberty.groupId}</groupId>
+                                    <artifactId>${liberty.artifactId}</artifactId>
+                                    <version>${version.openliberty-runtime}</version>
+                                    <type>zip</type>
+                                </assemblyArtifact>
```
Comment (yeekangc) [inline incubator/java-openliberty/image/project/pom.xml]: Do we need the assemblyArtifact?

Comment (scottkurz) [reply incubator/java-openliberty/image/project/pom.xml]: I think it's helpful to install this in the base, stack image so the user doesn't have to wait as long on `appsody run` to install it.

Comment (yeekangc) [reply incubator/java-openliberty/image/project/pom.xml]: Thanks, Scott.  We can update later to use the latest way to specify the runtime to install.

**Comment by yeekangc** [comment]
@neeraj-laad, we are ready for a 0.1 release.  Can you or someone on the team merge?

File: incubator/java-openliberty/README.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# Java OpenLiberty Stack
```
[the remark this answers is not in the document]
Comment (scottkurz) [reply incubator/java-openliberty/README.md]: Good point.. made this change.

**Comment by yeekangc** [comment]
I don't know who exactly has write access.  Tagging @kylegc @sympatheticmoose in case they can take a look and merge too.  Thank you.

File: incubator/java-openliberty/image/project/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+RUN cd /project/user-app && mvn liberty:create package
+
+RUN cd /project/user-app/target/liberty/wlp/usr/servers && \
+    mkdir /config && \
+    mv defaultServer/* /config/
+
+# Step 2: Package Open Liberty image
+FROM open-liberty:kernel-java8-openj9
+
+COPY --chown=1001:0 --from=0 /config/ /config/
```
[the remark this answers is not in the document]
Comment (scottkurz) [reply incubator/java-openliberty/image/project/Dockerfile]: Ah..thanks.

File: incubator/java-openliberty/image/project/pom.xml [hunk 4]
```diff
@@ -0,0 +1,123 @@
+<?xml version='1.0' encoding='utf-8'?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>dev.appsody</groupId>
+    <artifactId>java-openliberty</artifactId>
+    <version>0.1.0</version>
```
[the remark this answers is not in the document]
Comment (scottkurz) [reply incubator/java-openliberty/image/project/pom.xml]: Thanks...definitely want to do this soon, but was planning on using a later PR.

**Comment by scottkurz** [comment]
Thx @Kamran64 for the review.. hopefully my comments are clear enough, but since I rebased let me just note since the last PR, I just added the one new commit https://github.com/appsody/stacks/pull/573/commits/881396e698331be3d6ed1af9e3e3a3760db3e1d6.

File: incubator/java-openliberty/image/config/app-deploy.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+spec:
+  # Add fields here
+  version: 1.0.0
+  applicationImage: APPSODY_DOCKER_IMAGE 
+  stack: APPSODY_STACK
+  service:
+    type: NodePort
+    port: APPSODY_PORT
+    annotations:
+      prometheus.io/scrape: 'true'
```
[the remark this answers is not in the document]
Comment (scottkurz) [reply incubator/java-openliberty/image/config/app-deploy.yaml]: Thx for the suggestion.  Will give this some thought later on as we get more into metrics and monitoring.

**Comment by scottkurz** [comment]
Fixes #584, fixes #364

**Comment by skoh7645** [comment]
When comparing the time taken to execute run/build commands on the this `java-openliberty` stack vs the `java-microprofile` stack
Open Liberty
1. First appsody run - 43s
2. Recompile run on change - 13s
3. First build - 4m 4s

Microprofile
1. First appsody run - 1m 33s
2. Recompile run on change - 26s
3. First build - 5m 15s

**Comment by Kamran64** [comment]
@scottkurz I can access all endpoints except for javametrics-dash. Are you seeing this behaviour also?

**Comment by yeekangc** [comment]
Do we need javametrics still?  Codewind is moving away from that and will use standard metrics endpoints going forward?

**Pull request merged**